In [ ]:
#Name: Salman Mahboob
# No : 8813022

# Loading Packages

In [ ]:
import numpy as np
from sklearn.model_selection import train_test_split 
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_absolute_error , mean_absolute_percentage_error , r2_score
import matplotlib.pyplot as plt  
from sklearn.preprocessing import PolynomialFeatures

In [2]:
from sklearn import datasets
diabetes_X, diabetes_y = datasets.load_diabetes(return_X_y=True, as_frame=True, scaled=True)

# independent variables X
X = diabetes_X

# dependent variable y
y = diabetes_y

In [3]:
diabetes_X.head()

age       sex       bmi        bp        s1        s2        s3   
0  0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401  \
1 -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2  0.085299  0.050680  0.044451 -0.005670 -0.045599 -0.034194 -0.032356   
3 -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4  0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   

         s4        s5        s6  
0 -0.002592  0.019907 -0.017646  
1 -0.039493 -0.068332 -0.092204  
2 -0.002592  0.002861 -0.025930  
3  0.034309  0.022688 -0.009362  
4 -0.002592 -0.031988 -0.046641

In [ ]:
y_pred_all = model_all.predict(X_val)
y_pred_train_all = model_all.predict(X_train)


In [4]:
#To display few rows of diabetes_y 
print(diabetes_y.head())

0    151.0
1     75.0
2    141.0
3    206.0
4    135.0
Name: target, dtype: float64


In [5]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=1)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=1) 

In [6]:
# Train the model
model1 = LinearRegression()  
model1.fit(X_train, y_train)

LinearRegression()

In [7]:
y_val_pred1 =model1.predict(X_val)

In [8]:
print("Coefficients:",model1.coef_)
print("Intercept:",model1.intercept_)

# Extracting feature names
features = X_train.columns
feature_names1 = []
for feature in features:
    feature_names1.append(feature)

Coefficients: [  -5.13092568 -240.24959113  508.46066938  309.45184096 -700.12813937
  342.13116894   76.85497028  240.56839974  678.20477958  105.81379966]
Intercept: 147.8991730777127


In [9]:
X_train_bmi = np.array(X_train['bmi']).reshape(-1,1)
X_test_bmi = np.array(X_test['bmi']).reshape(-1,1)
X_val_bmi = np.array(X_val['bmi']).reshape(-1,1)

In [10]:
polynomial1 = PolynomialFeatures(degree=2, include_bias=False)
X_train_bmi_poly = polynomial1.fit_transform(X_train_bmi)
X_test_bmi_poly = polynomial1.transform(X_test_bmi)
X_val_bmi_poly = polynomial1.transform(X_val_bmi)

# Train the model
model2 = LinearRegression()
model2.fit(X_train_bmi_poly, y_train)

LinearRegression()

In [11]:
y_val_pred2 =model2.predict(X_val_bmi_poly)


In [12]:
# Print the coefficients and intercept
print("Coefficients:",model2.coef_)
print("Intercept:",model2.intercept_)

# Extract feature_names
feature_names2 = polynomial1.get_feature_names_out()

Coefficients: [ 920.5997892  1229.45505591]
Intercept: 147.07331894260696


In [13]:
# Polynomial regression model 
polynomial2 = PolynomialFeatures(degree=2, include_bias=False)
X_train_poly = polynomial2.fit_transform(X_train)
X_test_poly = polynomial2.transform(X_test)
X_val_poly = polynomial2.transform(X_val)

# Train the model
model3 = LinearRegression()
model3.fit(X_train_poly, y_train)

LinearRegression()

In [14]:
# Make predictions using the validation data
y_val_pred3 =model3.predict(X_val_poly)

In [15]:
# Print the coefficients and intercept
print("Coefficients:",model3.coef_)
print("Intercept:",model3.intercept_)

# Extract feature_names
feature_names3= polynomial2.get_feature_names_out()

Coefficients: [ 6.00691853e+01 -1.96503113e+02  5.25634029e+02  2.35889257e+02
 -5.76528486e+04  5.05755756e+04  2.12572071e+04  6.60406170e+00
  1.95369964e+04  6.47858949e+01  9.84100355e+02  3.00735114e+03
  1.77316098e+03 -2.37457301e+03 -2.35614424e+03 -1.99239911e+03
  6.09634193e+03  2.42782406e+03  4.44202534e+03  2.68805905e+03
 -1.18658061e+00  3.15004119e+03  3.39934741e+03  6.97947170e+03
 -8.78013801e+03 -5.01103893e+02  1.76683242e+03 -4.01250753e+03
 -1.99343408e+03  4.42551744e+02  5.28495216e+03  1.39648141e+04
 -1.23381122e+04 -4.41603106e+03 -3.57192070e+02 -5.13424529e+03
 -1.55886653e+03  3.78755403e+02 -3.48571711e+03  6.47121169e+03
  1.19326945e+03 -2.04172875e+03 -1.49780414e+03 -9.91285223e+02
  2.63764072e+04 -1.66118748e+04 -3.74251143e+04 -3.91165553e+04
  1.84494181e+05 -1.94007860e+04 -4.39392695e+03  1.48039753e+04
  1.98915240e+04 -1.63539236e+05  1.44391093e+04  1.40960592e+04
  2.99041963e+04 -7.17060520e+04  1.11793213e+04  1.41342133e+04
  1.0724807

In [16]:
# performance evaluation function
def evaluate_performance(y1,y2):
    print("Mean Absolute Percentage error:", round(mean_absolute_percentage_error(y1, y2)*100,2))
    print("Mean Absolute Error:",round(mean_absolute_error(y1, y2),2))
    print("R- Squared : ",round(r2_score(y1,y2),2))

In [17]:
print("----------------------PERFORMANCE EVALUATION ON VALIDATION DATA-------------------------------------")

print(" Multivariate linear regression ")
print("------------------------------------------------------------------------")
evaluate_performance(y_val,y_val_pred1)

print("Polynomial regression of the 2nd degree on the BMI feature alone : ")
print("------------------------------------------------------------------------")
evaluate_performance(y_val,y_val_pred2)

print("Multivariate polynomial regression of the 2nd degree on all variables : ")
print("-------------------------------------------------------------------------")
evaluate_performance(y_val,y_val_pred3)

----------------------PERFORMANCE EVALUATION ON VALIDATION DATA-------------------------------------
 Multivariate linear regression 
------------------------------------------------------------------------
Mean Absolute Percentage error: 32.98
Mean Absolute Error: 45.4
R- Squared :  0.51
Polynomial regression of the 2nd degree on the BMI feature alone : 
------------------------------------------------------------------------
Mean Absolute Percentage error: 42.47
Mean Absolute Error: 53.65
R- Squared :  0.36
Multivariate polynomial regression of the 2nd degree on all variables : 
-------------------------------------------------------------------------
Mean Absolute Percentage error: 32.59
Mean Absolute Error: 46.81
R- Squared :  0.46


For a Non-Expert
MAPE: Lower is better.

MAE: Smaller is better.

R2: Closer to 1 is better.

MY INSIGHTS:
Multivariate Linear Regression has the lowest Mean Absolute Percentage Error (MAPE=32.98%) and Mean Absolute Error (MAE=45.4) among the three models, indicating good predictive accuracy on the validation data. It is also performing the best, with a 51% improvement over a naive model.

Polynomial Regression (2nd Degree) on BMI Feature Alone has higher MAPE(42.47%) and MAE(53.65) compared to the multivariate linear regression model. It is still performing better than a naive model, with a 36% improvement.

Multivariate Polynomial Regression (2nd Degree) on All Variables, is slightly less accurate than the multivariate linear regression model. It performs better than a naive model (46% improvement).

How many parameters are we fitting for each of the three models?

In [18]:
num_parameters_model1 = len(model1.coef_) + 1
print("Number of Parameters For Multivariate Linear Regression:", num_parameters_model1)
print(feature_names1," and target")

Number of Parameters For Multivariate Linear Regression: 11
['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']  and target


In [19]:
num_parameters_model2 = len(model2.coef_) + 1
print("Number of Parameters for Polynomial Regression on BMI Feature:", num_parameters_model2)
print(feature_names2," (x0 means bmi and x0^2 means square of bmi) and target")

Number of Parameters for Polynomial Regression on BMI Feature: 3
['x0' 'x0^2']  (x0 means bmi and x0^2 means square of bmi) and target


In [20]:
num_parameters_model3 = len(model3.coef_) + 1
print("Number of Parameters for Multivariate Polynomial Regression:", num_parameters_model3)
print(feature_names3,"and target")

Number of Parameters for Multivariate Polynomial Regression: 66
['age' 'sex' 'bmi' 'bp' 's1' 's2' 's3' 's4' 's5' 's6' 'age^2' 'age sex'
 'age bmi' 'age bp' 'age s1' 'age s2' 'age s3' 'age s4' 'age s5' 'age s6'
 'sex^2' 'sex bmi' 'sex bp' 'sex s1' 'sex s2' 'sex s3' 'sex s4' 'sex s5'
 'sex s6' 'bmi^2' 'bmi bp' 'bmi s1' 'bmi s2' 'bmi s3' 'bmi s4' 'bmi s5'
 'bmi s6' 'bp^2' 'bp s1' 'bp s2' 'bp s3' 'bp s4' 'bp s5' 'bp s6' 's1^2'
 's1 s2' 's1 s3' 's1 s4' 's1 s5' 's1 s6' 's2^2' 's2 s3' 's2 s4' 's2 s5'
 's2 s6' 's3^2' 's3 s4' 's3 s5' 's3 s6' 's4^2' 's4 s5' 's4 s6' 's5^2'
 's5 s6' 's6^2'] and target


Which model would you choose for deployment, and why?
Training and validation completed, before deployment testing must be performed.
Based on Validation:
I will choose Multivariate linear regression model
In the validation test, multivariate linear regression model performed well.
Multivariate Linear Regression model has the lowest MAPE and MAE, and the highest R2 score.
Therefore, the Multivariate Linear Regression model would be the best choice for deployment.
It offers the best balance between prediction accuracy and goodness of fit to the data.
From the results below after Testing :
Multivariate linear regression model performed better than other models.

In [21]:
print("----------------------PREDICTION AND PERFORMANCE EVALUATION ON TEST DATA-------------------------------------")

print(" Multivariate linear regression ")
print("------------------------------------------------------------------------")
y_test_pred1 =model1.predict(X_test)
evaluate_performance(y_test,y_test_pred1)

print("Polynomial regression of the 2nd degree on the BMI feature alone : ")
print("------------------------------------------------------------------------")
y_test_pred2 =model2.predict(X_test_bmi_poly)
evaluate_performance(y_test,y_test_pred2)

print("Multivariate polynomial regression of the 2nd degree on all variables : ")
print("-------------------------------------------------------------------------")
y_test_pred3 =model3.predict(X_test_poly)
evaluate_performance(y_test,y_test_pred3)

----------------------PREDICTION AND PERFORMANCE EVALUATION ON TEST DATA-------------------------------------
 Multivariate linear regression 
------------------------------------------------------------------------
Mean Absolute Percentage error: 31.57
Mean Absolute Error: 41.53
R- Squared :  0.44
Polynomial regression of the 2nd degree on the BMI feature alone : 
------------------------------------------------------------------------
Mean Absolute Percentage error: 42.24
Mean Absolute Error: 50.16
R- Squared :  0.26
Multivariate polynomial regression of the 2nd degree on all variables : 
-------------------------------------------------------------------------
Mean Absolute Percentage error: 34.51
Mean Absolute Error: 46.8
R- Squared :  0.26


I will choose "Multivariate linear regression" Model among the three models.